### ISSUE TITLE
helloworld pod start failed

### ISSUE THREAD

**Issue by zchunhai** [alibaba/inclavare-containers#1068]
I test helloworld as the following two documents：

https://github.com/alibaba/inclavare-containers/blob/master/docs/create_a_confidential_computing_kubernetes_cluster_with_inclavare_containers.md

https://github.com/alibaba/inclavare-containers/blob/master/docs/develop_and_deploy_hello_world_application_in_kubernetes_cluster.md

## Versions

SGX driver: 2.11

linux-sgx: 2.9.101.2
(In order to find the cause of the problem, I added 2 lines log)

inclavare-containers: v0.6.0

## helloworld pod CrashLoopBackOff

root@sgx:~# kubectl describe pod helloworld

![1621841050966](https://user-images.githubusercontent.com/14309137/119314114-ece5ac80-bca6-11eb-91c5-2443abe5a588.jpg)

## sgx_sign gendata successful in host server

```
root@sgx:/tmp# /opt/intel/sgxsdk/bin/x64/sgx_sign gendata -enclave ./occlum_context/build/lib/libocclum-libos.so -config ./occlum_context/build/Enclave.xml -out enclave_sig.dat
<EnclaveConfiguration>
    <ProdID>0</ProdID>
    <ISVSVN>0</ISVSVN>
    <StackMinSize>1048576</StackMinSize>
    <StackMaxSize>1048576</StackMaxSize>
    <HeapMaxSize>33554432</HeapMaxSize>
    <HeapMinSize>33554432</HeapMinSize>
    <TCSNum>32</TCSNum>
    <TCSPolicy>1</TCSPolicy>
    <DisableDebug>0</DisableDebug>
    <MiscSelect>0</MiscSelect>
    <MiscMask>0xFFFFFFFF</MiscMask>
    <ReservedMemMaxSize>268435456</ReservedMemMaxSize>
    <ReservedMemMinSize>268435456</ReservedMemMinSize>
    <ReservedMemInitSize>268435456</ReservedMemInitSize>
    <ReservedMemExecutable>1</ReservedMemExecutable>
</EnclaveConfiguration>
tcs_num 32, tcs_max_num 32, tcs_min_pool 1
gdata -> sdk_version: 589925
VERSION_UINT: 589925
The required memory is 491339776B.
The required memory is 0x1d494000, 479824 KB.
Succeed.
```

**Comment by jiazhang0** [comment]
@zchunhai The root cause is due to the mismatched SGX SDK used for enclave build and signing.

The signing tool will do this check to confirm the enclave building and signing processes use the same version of SGX SDK:
```C
    // Check whether the same SDK is used for enclave building and signing
    global_data_t *gdata = (global_data_t *)get_rawdata_by_rva(rva);
    if (gdata -> sdk_version != VERSION_UINT)
    {
        se_trace(SE_TRACE_ERROR, SDK_VERSION_ERROR);
      return false;
    }
```

In your context, the signing tool is built based on SGX SDK 2.9.101, and the enclave binary is built based on SGX SDK 2.11.100.

**Comment by hustliyilin** [comment]
Thanks for your issue @zchunhai . In [create-a-pod-with-occlum-sdk-image](https://github.com/alibaba/inclavare-containers/blob/master/docs/develop_and_deploy_hello_world_application_in_kubernetes_cluster.md#1-create-a-pod-with-occlum-sdk-image), we can see we will use the `docker.io/occlum/occlum:0.21.0-centos8.2` docker image  and the docker images has a higher sgxsdk version(`2.13`).  So you need to choose `linux-sgx: 2.13` rather than `linux-sgx: 2.09 `. And @jiazhiguang  will update the docs as soon as possible, thanks again.

By the way, you can get sgxsdk version of `docker.io/occlum/occlum:0.21.0-centos8.2` through the following steps:
- docker run: `sudo docker run -it --rm --privileged occlum/occlum:0.21.0-centos8.2`
- check the sgxsdk version in the container: `/opt/intel/sgxsdk/bin/x64/sgx_sign -version`

![image](https://user-images.githubusercontent.com/6824093/119320792-8795b980-bcae-11eb-8bb6-4d3982721708.png)

**Comment by zchunhai** [comment]
Thanks for your reply @hustliyilin @jiazhang0 .
I installed the SDK 2.13.100.4 on the host as required, deployed the occlum builder with version 2.13.100.4, and rebuilt helloworld image, but the problem same.

```
root@sgx:~# kubectl exec -it occlum-app-builder -- /opt/intel/sgxsdk/bin/x64/sgx_sign -version

This is sgx_sign from Intel(R) Software Guard Extensions, version 2.13.100.4 for Linux.

Copyright (C) 2021 Intel Corporation

BSD License. The License file could be found under the Intel(R) SGX SDK package.

root@sgx:~# /opt/intel/sgxsdk/bin/x64/sgx_sign -version

This is sgx_sign from Intel(R) Software Guard Extensions, version 2.13.100.4 for Linux.

Copyright (C) 2021 Intel Corporation

BSD License. The License file could be found under the Intel(R) SGX SDK package.

root@sgx:~# kubectl describe pod helloworld
......
Events:
  Type     Reason     Age                   From               Message
  ----     ------     ----                  ----               -------
  Normal   Scheduled  <unknown>             default-scheduler  Successfully assigned default/helloworld to sgx
  Normal   Pulled     7m1s (x5 over 8m29s)  kubelet, sgx       Container image "clustarai/occlum-hello-world:v2" already present on machine
  Normal   Created    7m1s (x5 over 8m29s)  kubelet, sgx       Created container helloworld
  Warning  Failed     7m1s (x5 over 8m28s)  kubelet, sgx       Error: failed to create containerd task: +++ dirname /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/data/carrier.sh
++ cd /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/data
++ pwd -P
+ base_dir=/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/data
+ occlum_workspace=/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/data/occlum_workspace
++ getopt -a -o a:r:w:p:c:e:u:m:s:k:n: -l action:,rootfs:,work_dir:,entry_point:,occlum_config_path:,enclave_config_path:,unsigned_enclave_path:,unsigned_material_path:,signed_enclave_path:,public_key_path:,signature_path: -- --action generateSigningMaterial --enclave_config_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml --unsigned_enclave_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so --unsigned_material_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat
+ temp=' --action '\''generateSigningMaterial'\'' --enclave_config_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml'\'' --unsigned_enclave_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so'\'' --unsigned_material_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat'\'' --'
+ eval set -- ' --action '\''generateSigningMaterial'\'' --enclave_config_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml'\'' --unsigned_enclave_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so'\'' --unsigned_material_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat'\'' --'
++ set -- --action generateSigningMaterial --enclave_config_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml --unsigned_enclave_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so --unsigned_material_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat --
+ true
+ case "$1" in
+ action=generateSigningMaterial
+ shift
+ shift
+ true
+ case "$1" in
+ enclave_config_path=/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml
+ shift
+ shift
+ true
+ case "$1" in
+ unsigned_enclave_path=/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so
+ shift
+ shift
+ true
+ case "$1" in
+ unsigned_material_path=/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat
+ shift
+ shift
+ true
+ case "$1" in
+ shift
+ break
+ doAction
+ '[' generateSigningMaterial == '' ']'
+ case ${action} in
+ generateSigningMaterial
+ [[ /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml == '' ]]
+ [[ /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so == '' ]]
+ [[ /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat == '' ]]
+ /opt/intel/sgxsdk/bin/x64/sgx_sign gendata -enclave /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so -config /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml -out /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat
tcs_num 32, tcs_max_num 32, tcs_min_pool 1
SDK version is not correct. The same SDK should be used for enclave building and signing.
The SDK version for building enclave could be obtained by below command:

           $ strings {Enclave.so} | grep SGX_TSTDC_VERSION

Error happened while signing the enclave.
<EnclaveConfiguration>
    <ProdID>0</ProdID>
    <ISVSVN>0</ISVSVN>
    <StackMaxSize>1048576</StackMaxSize>
    <StackMinSize>1048576</StackMinSize>
    <HeapMaxSize>33554432</HeapMaxSize>
    <HeapMinSize>33554432</HeapMinSize>
    <TCSNum>32</TCSNum>
    <TCSPolicy>1</TCSPolicy>
    <DisableDebug>0</DisableDebug>
    <MiscSelect>0</MiscSelect>
    <MiscMask>0xFFFFFFFF</MiscMask>
    <ReservedMemMaxSize>268435456</ReservedMemMaxSize>
    <ReservedMemMinSize>268435456</ReservedMemMinSize>
    <ReservedMemInitSize>268435456</ReservedMemInitSize>
    <ReservedMemExecutable>1</ReservedMemExecutable>
</EnclaveConfiguration>
 exit status 255: unknown
  Warning  BackOff  3m16s (x24 over 8m26s)  kubelet, sgx  Back-off restarting failed container
root@sgx:~#
```

**Comment by hustliyilin** [comment]
> Thanks for your reply @hustliyilin @jiazhang0 .
> I installed the SDK 2.13.100.4 on the host as required, deployed the occlum builder with version 2.13.100.4, and rebuilt helloworld image, but the problem same.
> 
> ```
> root@sgx:~# kubectl exec -it occlum-app-builder -- /opt/intel/sgxsdk/bin/x64/sgx_sign -version
> 
> This is sgx_sign from Intel(R) Software Guard Extensions, version 2.13.100.4 for Linux.
> 
> Copyright (C) 2021 Intel Corporation
> 
> BSD License. The License file could be found under the Intel(R) SGX SDK package.
> 
> root@sgx:~# /opt/intel/sgxsdk/bin/x64/sgx_sign -version
> 
> This is sgx_sign from Intel(R) Software Guard Extensions, version 2.13.100.4 for Linux.
> 
> Copyright (C) 2021 Intel Corporation
> 
> BSD License. The License file could be found under the Intel(R) SGX SDK package.
> 
> root@sgx:~# kubectl describe pod helloworld
> ......
> Events:
>   Type     Reason     Age                   From               Message
>   ----     ------     ----                  ----               -------
>   Normal   Scheduled  <unknown>             default-scheduler  Successfully assigned default/helloworld to sgx
>   Normal   Pulled     7m1s (x5 over 8m29s)  kubelet, sgx       Container image "clustarai/occlum-hello-world:v2" already present on machine
>   Normal   Created    7m1s (x5 over 8m29s)  kubelet, sgx       Created container helloworld
>   Warning  Failed     7m1s (x5 over 8m28s)  kubelet, sgx       Error: failed to create containerd task: +++ dirname /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/data/carrier.sh
> ++ cd /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/data
> ++ pwd -P
> + base_dir=/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/data
> + occlum_workspace=/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/data/occlum_workspace
> ++ getopt -a -o a:r:w:p:c:e:u:m:s:k:n: -l action:,rootfs:,work_dir:,entry_point:,occlum_config_path:,enclave_config_path:,unsigned_enclave_path:,unsigned_material_path:,signed_enclave_path:,public_key_path:,signature_path: -- --action generateSigningMaterial --enclave_config_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml --unsigned_enclave_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so --unsigned_material_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat
> + temp=' --action '\''generateSigningMaterial'\'' --enclave_config_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml'\'' --unsigned_enclave_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so'\'' --unsigned_material_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat'\'' --'
> + eval set -- ' --action '\''generateSigningMaterial'\'' --enclave_config_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml'\'' --unsigned_enclave_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so'\'' --unsigned_material_path '\''/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat'\'' --'
> ++ set -- --action generateSigningMaterial --enclave_config_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml --unsigned_enclave_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so --unsigned_material_path /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat --
> + true
> + case "$1" in
> + action=generateSigningMaterial
> + shift
> + shift
> + true
> + case "$1" in
> + enclave_config_path=/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml
> + shift
> + shift
> + true
> + case "$1" in
> + unsigned_enclave_path=/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so
> + shift
> + shift
> + true
> + case "$1" in
> + unsigned_material_path=/run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat
> + shift
> + shift
> + true
> + case "$1" in
> + shift
> + break
> + doAction
> + '[' generateSigningMaterial == '' ']'
> + case ${action} in
> + generateSigningMaterial
> + [[ /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml == '' ]]
> + [[ /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so == '' ]]
> + [[ /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat == '' ]]
> + /opt/intel/sgxsdk/bin/x64/sgx_sign gendata -enclave /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/lib/libocclum-libos.so -config /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/build/Enclave.xml -out /run/containerd/io.containerd.runtime.v2.task/k8s.io/helloworld/rootfs/var/run/rune/enclave_sig.dat
> tcs_num 32, tcs_max_num 32, tcs_min_pool 1
> SDK version is not correct. The same SDK should be used for enclave building and signing.
> The SDK version for building enclave could be obtained by below command:
> 
>            $ strings {Enclave.so} | grep SGX_TSTDC_VERSION
> 
> Error happened while signing the enclave.
> <EnclaveConfiguration>
>     <ProdID>0</ProdID>
>     <ISVSVN>0</ISVSVN>
>     <StackMaxSize>1048576</StackMaxSize>
>     <StackMinSize>1048576</StackMinSize>
>     <HeapMaxSize>33554432</HeapMaxSize>
>     <HeapMinSize>33554432</HeapMinSize>
>     <TCSNum>32</TCSNum>
>     <TCSPolicy>1</TCSPolicy>
>     <DisableDebug>0</DisableDebug>
>     <MiscSelect>0</MiscSelect>
>     <MiscMask>0xFFFFFFFF</MiscMask>
>     <ReservedMemMaxSize>268435456</ReservedMemMaxSize>
>     <ReservedMemMinSize>268435456</ReservedMemMinSize>
>     <ReservedMemInitSize>268435456</ReservedMemInitSize>
>     <ReservedMemExecutable>1</ReservedMemExecutable>
> </EnclaveConfiguration>
>  exit status 255: unknown
>   Warning  BackOff  3m16s (x24 over 8m26s)  kubelet, sgx  Back-off restarting failed container
> root@sgx:~#
> ```

@zchunhai What's your host sgxsdk version? you should check the sgxsdk version in the host through `/opt/intel/sgxsdk/bin/x64/sgx_sign -version`

By the way, the URL you provided is the newest user guide rather than v0.6.0
- https://github.com/alibaba/inclavare-containers/blob/master/docs/create_a_confidential_computing_kubernetes_cluster_with_inclavare_containers.md
- https://github.com/alibaba/inclavare-containers/blob/master/docs/develop_and_deploy_hello_world_application_in_kubernetes_cluster.md

And you say you use inclavare containers v0.6.0, it means you should user the corresponding URL, Thanks again

- https://github.com/alibaba/inclavare-containers/blob/v0.6.0/docs/create_a_confidential_computing_kubernetes_cluster_with_inclavare_containers.md 
- https://github.com/alibaba/inclavare-containers/blob/v0.6.0/docs/develop_and_deploy_hello_world_application_in_kubernetes_cluster.md

**Comment by jiazhang0** [comment]
@zchunhai Assuming using sgx sdk 2.13, what is this occlum app image clustarai/occlum-hello-world:v2 built on? If it was built with occlum image 0.17.0, you have to rebuild it in occlum image 0.21.0.

**Comment by jiazhang0** [comment]
In addition, please show us the version of shim installed. If

**Comment by zchunhai** [comment]
Thanks @jiazhang0 ! Now all versions are correct, as the description of release 0.6.0. But there is  a new error with  the helloworld pod:

```
root@sgx:~# kubectl describe pod helloworld1
Name:         helloworld1
Namespace:    default
Priority:     0
Node:         sgx/192.168.1.41
Start Time:   Tue, 25 May 2021 15:33:30 +0800
Labels:       run=helloworld
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"labels":{"run":"helloworld"},"name":"helloworld1","namespace":"default"},"sp...
Status:       Running
IP:           172.21.0.220
IPs:
  IP:  172.21.0.220
Containers:
  helloworld:
    Container ID:  containerd://a438bd84617895a1b6eb34f56b0e7caa5a956708db84a6a94118119648972e4d
    Image:         clustarai/occlum-hello-world:v3
    Image ID:      docker.io/clustarai/occlum-hello-world@sha256:35c0d817f0ba5ac95978c80707fda6674045511267c1f49bb5ed8e149440937a
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/hello_world
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       StartError
      Message:      failed to create containerd task: ttrpc: closed: unknown
      Exit Code:    128
      Started:      Thu, 01 Jan 1970 08:00:00 +0800
      Finished:     Tue, 25 May 2021 17:02:08 +0800
    Ready:          False
    Restart Count:  22
    Environment:
      RUNE_CARRIER:  occlum
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-vlkzt (ro)
......
```

I can run this image directly on the host.
```
root@sgx:~# ctr run docker.io/clustarai/occlum-hello-world:v3 v3
Hello World!
Hello World!
Hello World!
Hello World!
root@sgx:~#
```

**Comment by hustliyilin** [comment]
> Thanks @jiazhang0 ! Now all versions are correct, as the description of release 0.6.0. But there is a new error with the helloworld pod:
> 
> ```
> root@sgx:~# kubectl describe pod helloworld1
> Name:         helloworld1
> Namespace:    default
> Priority:     0
> Node:         sgx/192.168.1.41
> Start Time:   Tue, 25 May 2021 15:33:30 +0800
> Labels:       run=helloworld
> Annotations:  kubectl.kubernetes.io/last-applied-configuration:
>                 {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"labels":{"run":"helloworld"},"name":"helloworld1","namespace":"default"},"sp...
> Status:       Running
> IP:           172.21.0.220
> IPs:
>   IP:  172.21.0.220
> Containers:
>   helloworld:
>     Container ID:  containerd://a438bd84617895a1b6eb34f56b0e7caa5a956708db84a6a94118119648972e4d
>     Image:         clustarai/occlum-hello-world:v3
>     Image ID:      docker.io/clustarai/occlum-hello-world@sha256:35c0d817f0ba5ac95978c80707fda6674045511267c1f49bb5ed8e149440937a
>     Port:          <none>
>     Host Port:     <none>
>     Command:
>       /bin/hello_world
>     State:          Waiting
>       Reason:       CrashLoopBackOff
>     Last State:     Terminated
>       Reason:       StartError
>       Message:      failed to create containerd task: ttrpc: closed: unknown
>       Exit Code:    128
>       Started:      Thu, 01 Jan 1970 08:00:00 +0800
>       Finished:     Tue, 25 May 2021 17:02:08 +0800
>     Ready:          False
>     Restart Count:  22
>     Environment:
>       RUNE_CARRIER:  occlum
>     Mounts:
>       /var/run/secrets/kubernetes.io/serviceaccount from default-token-vlkzt (ro)
> ......
> ```
> 
> I can run this image directly on the host.
> 
> ```
> root@sgx:~# ctr run docker.io/clustarai/occlum-hello-world:v3 v3
> Hello World!
> Hello World!
> Hello World!
> Hello World!
> root@sgx:~#
> ```

Inclavare Containers released the v0.6.0 version on February 10, 2021. Occlum supports the release of sgxsdk 2.13 in March 2021, and Inclavare also supports sgxsdk 2.13 in 2021. Therefore, if you want to use sgxsdk 2.13, it is best to use the latest inclavare container. I’m happy to tell you that we will release v0.6.1 at the end of this month. @zchunhai  In the future, we mainly support sgxsdk 2.13 and occlum 0.21.0. 

![image](https://user-images.githubusercontent.com/6824093/119489878-bedb9780-bd8e-11eb-997f-2b2e711b676d.png)

![image](https://user-images.githubusercontent.com/6824093/119489983-dca8fc80-bd8e-11eb-97a6-e4ac3a531d42.png)

If you still want to use v0.6.0, you should use sgxsdk 2.11 and you can refer to https://github.com/alibaba/inclavare-containers/blob/v0.6.0/docs/create_a_confidential_computing_kubernetes_cluster_with_inclavare_containers.md and
https://github.com/alibaba/inclavare-containers/blob/v0.6.0/docs/develop_and_deploy_hello_world_application_in_kubernetes_cluster.md

**Comment by jiazhang0** [comment]
@haosanzi Please follow the doc to confirm whether this issue still exists in 0.6.2.
